bson-codec-design/DESIGN.md

```diff
@@ -153,6 +162,21 @@ Unmarshal functions handle transforming to and from a slice of bytes while the E
 handle transforming to an `io.Writer` and from an `io.Reader`. These naming conventions follow the
 `encoding` subpackages of the standard library.
 
+#### Encoder and Decoder
+The `Encoder` and `Decoder` types can serialize and unserialize both BSON and Extended JSON. The
+serialization type is determined by a mode specified when creating the `Encoder` or `Decoder`.
+
+The serialization mode and `io.Reader`/`io.Writer` can be updated on an `Decoder` or `Encoder` by
+using the `Reset` method. This allows these types to be pooled and the user to configure the type
+when they use it.
+
+#### DocumentEncoder and DocumentDecoder
```

Comment:
I'm not sure why we need different types for these? It feels like an Encoder/Decoder would server this purpose and you are just trying to make it easier for a user in which case these could simply be functions which setup an Encoder/Decoder to use the document implementation of a ValueWriter/ValueReader.